x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,342 @@
+                    inferenceResultsSupplier.set(inferenceResponse.getResults());
+                    l.onResponse(null);
+                })
+            )
+        );
+
+        return new SemanticQueryBuilder(this, inferenceResultsSupplier);
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if (inferenceResults == null) {
+            throw new IllegalStateException("Query builder must be rewritten first");
+        }
+
+        List<? extends InferenceResults> inferenceResultsList = inferenceResults.transformToCoordinationFormat();
+        if (inferenceResultsList.isEmpty()) {
+            throw new IllegalArgumentException("No inference results retrieved for field [" + fieldName + "]");
+        } else if (inferenceResultsList.size() > 1) {
+            // TODO: How to handle multiple inference results?
+            throw new IllegalArgumentException(inferenceResultsList.size() + " inference results retrieved for field [" + fieldName + "]");
+        }
+
+        InferenceResults inferenceResults = inferenceResultsList.get(0);
+        MappedFieldType fieldType = context.getFieldType(fieldName);
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType == false) {
+            // TODO: Better exception type to throw here?
+            throw new IllegalArgumentException(
+                "Field [" + fieldName + "] is not registered as a " + SemanticTextFieldMapper.CONTENT_TYPE + " field type"
+            );
+        }
+
+        return semanticQuery(inferenceResults, context);
+    }
+
+    private Map<String, Set<String>> computeModelsForFields(Collection<IndexMetadata> indexMetadataCollection) {
+        Map<String, Set<String>> modelsForFields = new HashMap<>();
+        for (IndexMetadata indexMetadata : indexMetadataCollection) {
+            Map<String, Set<String>> fieldsForModels = indexMetadata.getFieldsForModels();
+            for (Map.Entry<String, Set<String>> entry : fieldsForModels.entrySet()) {
+                for (String fieldName : entry.getValue()) {
+                    Set<String> models = modelsForFields.computeIfAbsent(fieldName, v -> new HashSet<>());
+                    models.add(entry.getKey());
+                }
+            }
+        }
+
+        return modelsForFields;
+    }
+
+    private Query semanticQuery(InferenceResults inferenceResults, SearchExecutionContext context) {
+        // Cant use QueryBuilders.boolQuery() because a mapper is not registered for <field>.inference, causing
+        // TermQueryBuilder#doToQuery to fail (at TermQueryBuilder:202)
+        // TODO: Handle boost and queryName
+        String inferenceResultsFieldName = fieldName + "." + INFERENCE_CHUNKS_RESULTS;
+        BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder().setMinimumNumberShouldMatch(1);
+
+        // TODO: Support dense vectors
+        if (inferenceResults instanceof TextExpansionResults textExpansionResults) {
```

Comment:
This is one of the reasons I'd like this code to be moved to the FieldMapper - this class deals with a lot of internal details about the mapping structure otherwise.

Reply:
If that's the case, maybe `SemanticTextInferenceResultFieldMapper` is a better place for this method then? That's where all the mapping structure implementation is.